### PR TITLE
Implement Case-insensitive SNI matching

### LINKED ISSUE #4081: Multiple wildcard certificates, correct subdomain wildcard no longer selected
### Do you want to request a *feature* or report a *bug*?
Bug

### What did you do?
I upgraded from 1.5.3 to 1.7.3.  Traefik seems to select and use the wrong certificates now.  I have mutiple wildcard certificates in the likes of:
- *.domain.com
- *.int.domain.com
- *.app.domain.com (edited)
Now all sub domain hostnames such as test.int.domain.com are using the *.domain.com certificate

### What did you expect to see?
Similar behavior as 1.5 up to 1.6.6, the wildcard being served matching the proper wildcard subdomain certificate

### What did you see instead?
 all sub domain hostnames such as test.int.domain.com are using the *.domain.com certificate

### Output of `traefik version`: (_What version of Traefik are you using?_)
Traefik 1.7.3

### What is your environment & configuration (arguments, toml, provider, platform, ...)?
```
arguments
        - -d
        - --kubernetes
        - --consul
        - --consul.endpoint=consul-consul.infra.svc.cluster.local:8500
```
Certificates are configured in Consul KV and confirmed working in earlier versions.  

### If applicable, please paste the log output in DEBUG level (`--logLevel=DEBUG` switch)
```
time="2018-10-19T09:00:58Z" level=info msg="No tls.defaultCertificate given for https: using the first item in tls.certificates as a fallback."
time="2018-10-19T09:00:58Z" level=info msg="Traefik version v1.7.3 built on 2018-10-15_10:13:00AM"
time="2018-10-19T09:00:58Z" level=debug msg="Global configuration loaded  ... [[ removed because of sensitive data... ]]
time="2018-10-19T09:00:58Z" level=info msg="\nStats collection is disabled.\nHelp us improve Traefik by turning this feature on :)\nMore details on: https://docs.traefik.io/basics/#collected-data\n"
time="2018-10-19T09:00:58Z" level=debug msg="Configured Prometheus metrics"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.domain.com,domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.int.domain.com,int.domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.app.domain.com,app.domain.com"
time="2018-10-19T09:00:58Z" level=info msg="Preparing server http &{Address::80 TLS:<nil> Redirect:<nil> Auth:<nil> WhitelistSourceRange:[] WhiteList:<nil> Compress:false ProxyProtocol:<nil> ForwardedHeaders:0xc000370b80} with readTimeout=0s writeTimeout=0s idleTimeout=3m0s"
time="2018-10-19T09:00:58Z" level=info msg="Preparing server https &{Address::443 TLS:0xc0002a4ab0 Redirect:<nil> Auth:<nil> WhitelistSourceRange:[] WhiteList:<nil> Compress:false ProxyProtocol:<nil> ForwardedHeaders:0xc000370bc0} with readTimeout=0s writeTimeout=0s idleTimeout=3m0s"
time="2018-10-19T09:00:58Z" level=info msg="Starting server on :80"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.domain.com,domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.int.domain.com,int.domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.app.domain.com,app.domain.com"
time="2018-10-19T09:00:58Z" level=info msg="Preparing server traefik &{Address::8181 TLS:<nil> Redirect:<nil> Auth:<nil> WhitelistSourceRange:[] WhiteList:<nil> Compress:false ProxyProtocol:<nil> ForwardedHeaders:0xc000370b60} with readTimeout=0s writeTimeout=0s idleTimeout=3m0s"
time="2018-10-19T09:00:58Z" level=info msg="Starting server on :443"
time="2018-10-19T09:00:58Z" level=info msg="Starting provider configuration.ProviderAggregator {}"
time="2018-10-19T09:00:58Z" level=info msg="Starting server on :8181"
time="2018-10-19T09:00:58Z" level=debug msg="Node cb3a601c-9579-4b40-aa0c-6c65bde806de running for election"
time="2018-10-19T09:00:58Z" level=info msg="Starting provider *consul.Provider {\"Watch\":true,\"Filename\":\"\",\"Constraints\":[],\"Trace\":false,\"TemplateVersion\":0,\"DebugLogGeneratedTemplate\":false,\"Endpoint\":\"consul-consul.infra.svc.cluster.local:8500\",\"Prefix\":\"traefik\",\"TLS\":null,\"Username\":\"\",\"Password\":\"\"}"
time="2018-10-19T09:00:58Z" level=info msg="Starting provider *kubernetes.Provider {\"Watch\":true,\"Filename\":\"\",\"Constraints\":[],\"Trace\":false,\"TemplateVersion\":0,\"DebugLogGeneratedTemplate\":false,\"Endpoint\":\"\",\"Token\":\"\",\"CertAuthFilePath\":\"\",\"DisablePassHostHeaders\":false,\"EnablePassTLSCert\":false,\"Namespaces\":null,\"LabelSelector\":\"\",\"IngressClass\":\"\",\"IngressEndpoint\":null}"
time="2018-10-19T09:00:58Z" level=debug msg="Using Ingress label selector: \"\""
time="2018-10-19T09:00:58Z" level=info msg="ingress label selector is: \"\""
time="2018-10-19T09:00:58Z" level=info msg="Creating in-cluster Provider client"
time="2018-10-19T09:00:58Z" level=info msg="Node cb3a601c-9579-4b40-aa0c-6c65bde806de elected worker ♝"
time="2018-10-19T09:00:58Z" level=debug msg="Datastore reload"
time="2018-10-19T09:00:58Z" level=debug msg="Building ACME client..."
time="2018-10-19T09:00:58Z" level=error msg="Cannot unmarshall private key []"
time="2018-10-19T09:00:58Z" level=error msg="Error building ACME client &{Email: Registration:<nil> PrivateKey:[] KeyType: DomainsCertificate:{Certs:[] lock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0}} ChallengeCerts:map[] HTTPChallenge:map[]}: private key was nil"
time="2018-10-19T09:00:58Z" level=debug msg="Cannot get key traefik/alias Key not found in store, setting default traefik"
time="2018-10-19T09:00:58Z" level=debug msg="Cannot list keys under \"traefik/backends/\": Key not found in store"
time="2018-10-19T09:00:58Z" level=debug msg="Cannot get key traefik/alias Key not found in store, setting default traefik"
time="2018-10-19T09:00:58Z" level=debug msg="Cannot list keys under \"traefik/frontends/\": Key not found in store"
time="2018-10-19T09:00:58Z" level=debug msg="Cannot list keys under \"traefik/tls/\": Key not found in store"
time="2018-10-19T09:00:58Z" level=debug msg="Configuration received from provider consul: {}"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.domain.com,domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Cannot list keys under \"traefik/backends/\": Key not found in store"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.int.domain.com,int.domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.app.domain.com,app.domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Cannot list keys under \"traefik/frontends/\": Key not found in store"
time="2018-10-19T09:00:58Z" level=info msg="Server configuration reloaded on :8181"
time="2018-10-19T09:00:58Z" level=info msg="Server configuration reloaded on :80"
time="2018-10-19T09:00:58Z" level=info msg="Server configuration reloaded on :443"
time="2018-10-19T09:00:58Z" level=debug msg="Cannot list keys under \"traefik/tls/\": Key not found in store"
time="2018-10-19T09:00:58Z" level=debug msg="Configuration received from provider consul: {}"
time="2018-10-19T09:00:58Z" level=info msg="Skipping same configuration for provider consul"
time="2018-10-19T09:00:58Z" level=debug msg="Received Kubernetes event kind *v1beta1.Ingress"
time="2018-10-19T09:00:58Z" level=debug msg="Configuration received from provider kubernetes: {\"backends\":{\"alertmanager.app.domain.com/\":{\"servers\":{\"alertmanager-main-0\":{\"url\":\"http://10.99.85.233:9093\",\"weight\":1},\"alertmanager-main-1\":{\"url\":\"http://10.99.72.137:9093\",\"weight\":1},\"alertmanager-main-2\":{\"url\":\"http://10.99.105.51:9093\",\"weight\":1}},\"loadBalancer\":{\"method\":\"wrr\"}},\"blackbox.app.domain.com/\":{\"servers\":{\"blackbox-exporter-6cd8cf6875-22gz6\":{\"url\":\"https://10.99.105.37:9115\",\"weight\":1}},   [[ Data removed ]]

time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.domain.com,domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.int.domain.com,int.domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Add certificate for domains *.app.domain.com,app.domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Wiring frontend alertmanager.app.domain.com/ to entryPoint http"
time="2018-10-19T09:00:58Z" level=debug msg="Creating backend alertmanager.app.domain.com/"
time="2018-10-19T09:00:58Z" level=debug msg="Adding TLSClientHeaders middleware for frontend alertmanager.app.domain.com/"
time="2018-10-19T09:00:58Z" level=debug msg="Creating load-balancer wrr"
time="2018-10-19T09:00:58Z" level=debug msg="Creating server alertmanager-main-2 at http://10.99.105.51:9093 with weight 1"
time="2018-10-19T09:00:58Z" level=debug msg="Creating server alertmanager-main-0 at http://10.99.85.233:9093 with weight 1"
time="2018-10-19T09:00:58Z" level=debug msg="Creating server alertmanager-main-1 at http://10.99.72.137:9093 with weight 1"
time="2018-10-19T09:00:58Z" level=debug msg="Creating route alertmanager.app.domain.com Host:alertmanager.app.domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Creating route / PathPrefix:/"
time="2018-10-19T09:00:58Z" level=debug msg="Wiring frontend alertmanager.app.domain.com/ to entryPoint https"
time="2018-10-19T09:00:58Z" level=debug msg="Creating backend alertmanager.app.domain.com/"
time="2018-10-19T09:00:58Z" level=debug msg="Adding TLSClientHeaders middleware for frontend alertmanager.app.domain.com/"
time="2018-10-19T09:00:58Z" level=debug msg="Creating load-balancer wrr"
time="2018-10-19T09:00:58Z" level=debug msg="Creating server alertmanager-main-1 at http://10.99.72.137:9093 with weight 1"
time="2018-10-19T09:00:58Z" level=debug msg="Creating server alertmanager-main-2 at http://10.99.105.51:9093 with weight 1"
time="2018-10-19T09:00:58Z" level=debug msg="Creating server alertmanager-main-0 at http://10.99.85.233:9093 with weight 1"
time="2018-10-19T09:00:58Z" level=debug msg="Creating route alertmanager.app.domain.com Host:alertmanager.app.domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Creating route / PathPrefix:/"
time="2018-10-19T09:00:58Z" level=debug msg="Wiring frontend blackbox.app.domain.com/ to entryPoint http"
time="2018-10-19T09:00:58Z" level=debug msg="Creating backend blackbox.app.domain.com/"
time="2018-10-19T09:00:58Z" level=debug msg="Adding TLSClientHeaders middleware for frontend blackbox.app.domain.com/"
time="2018-10-19T09:00:58Z" level=debug msg="Creating load-balancer wrr"
time="2018-10-19T09:00:58Z" level=debug msg="Creating server blackbox-exporter-6cd8cf6875-22gz6 at https://10.99.105.37:9115 with weight 1"
time="2018-10-19T09:00:58Z" level=debug msg="Creating route blackbox.app.domain.com Host:blackbox.app.domain.com"
time="2018-10-19T09:00:58Z" level=debug msg="Creating route / PathPrefix:/"
time="2018-10-19T09:00:58Z" level=debug msg="Wiring frontend blackbox.app.domain.com/ to entryPoint https"
time="2018-10-19T09:00:58Z" level=debug msg="Creating backend blackbox.app.domain.com/"
time="2018-10-19T09:00:58Z" level=debug msg="Adding TLSClientHeaders middleware for frontend blackbox.app.domain.com/"
time="2018-10-19T09:00:58Z" level=debug msg="Creating load-balancer wrr"
time="2018-10-19T09:00:58Z" level=debug msg="Creating server blackbox-exporter-6cd8cf6875-22gz6 at https://10.99.105.37:9115 with weight 1"
time="2018-10-19T09:00:58Z" level=debug msg="Creating route / PathPrefix:/"
time="2018-10-19T09:00:58Z" level=debug msg="Creating route blackbox.app.domain.com Host:blackbox.app.domain.com"
...
```

### PR DESCRIPTION
**Pull request by dtomcej**
### What does this PR do?

Implements case-insensitive SNI matching for certificate selection

### Motivation

If your certificates did not match with case sensitivity, the default certificate would be served (which may be the incorrect one).

Since the `types.CanonicalDomain(clientHello.ServerName)` always returns lowercase, we should have all our certs keyed in lowercase.

Addresses issues in #4081 

### More

- [x] Added/updated tests
- [x] Added/updated documentation - None needed, Bugfix

### CODE REVIEW

File: tls/certificate.go [hunk 1]
```diff
@@ -154,21 +154,21 @@ func (c *Certificate) AppendCertificates(certs map[string]map[string]*tls.Certif
 
 	var SANs []string
 	if parsedCert.Subject.CommonName != "" {
-		SANs = append(SANs, parsedCert.Subject.CommonName)
+		SANs = append(SANs, strings.ToLower(parsedCert.Subject.CommonName))
```
Comment (dtomcej) [inline tls/certificate.go]: TODO: Replace ToLower

Comment (emilevauge) [reply tls/certificate.go]: Self review :smile:

**Comment by juliens** [comment]
As far as I remember, with the default GetCertificate in go, if you use a certificate with uppercase letter, it will never matchs. This PR change this behavior? Do you know what RFCs says on this?

**Comment by maxmeyer** [comment]
@juliens According to https://security.stackexchange.com/a/150776 and the referenced RFCs using uppercase letters in "Common Names" is allowed.

**Pull request merged**